src/teleop/jetson/arm_trajectory_server.py

```diff
@@ -0,0 +1,143 @@
+            point.effort = temp_effort
+
+        joint_trajectory.joint_names = conf_joint_names
+
+    # Action initialisation
+    def __init__(self, name: str) -> None:
+        self.publisher = rospy.Publisher("ra_cmd", JointState, queue_size=100)
+        self._action_name = name
+        self._as = actionlib.SimpleActionServer(
+            self._action_name, FollowJointTrajectoryAction, execute_cb=self.execute_cb, auto_start=False
+        )
+        self.trajectory_point_queue = deque()
+        self.desired_joint_state = JointState(name=conf_joint_names)
+        # Publisher for gazebo position controller
+        self.gazebo_pub = rospy.Publisher("gazebo_arm_controller/command", Float64MultiArray, queue_size=100)
+        self._as.start()
+
+    # Action callback
+    def execute_cb(self, goal: FollowJointTrajectoryGoal) -> None:
+
+        rospy.loginfo("Executing FollowJointTrajectory Action")
+        # It is required to rearrange the arrays because MoveIt doesn't guarantee order preservation
+        self.rearrange(goal.trajectory)
+
+        # A trajectory needs at least 2 points
+        if len(goal.trajectory.points) < 2:
+            return
+
+        time_start = rospy.Time.from_sec(time.time())
+
+        # ------------- Send command list
+
+        for point in goal.trajectory.points:
+            self.trajectory_point_queue.append(point)
+
+        # ------------- Wait until the termination while providing feedback
+
+        last_point = self.trajectory_point_queue[0]
+        self.trajectory_point_queue.popleft()
+        #Initialize gazebo Float array
+        gazebo_positions = Float64MultiArray(data = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
+        while len(self.trajectory_point_queue) > 0:
+            point = self.trajectory_point_queue[0]
+            self.trajectory_point_queue.popleft()
+            self.desired_joint_state.position = point.positions
+            self.publisher.publish(self.desired_joint_state)
+            gazebo_positions.data = point.positions
+            self.gazebo_pub.publish(gazebo_positions)
+            # Wait
+            rospy.sleep(point.time_from_start - last_point.time_from_start)
+            # Trajectory abort!
+            # To abort the current movement, it is possible to send an empty trajectory
+            if self._as.is_preempt_requested():
+                self.publisher.publish(joint_states)
+                self._as.set_preempted()
+                return
+            # ---------------------------------------
+            # Feedback
+            self._feedback.joint_names = goal.trajectory.joint_names
+            self._feedback.desired = point
+            with lock:
+                self._feedback.actual.positions = joint_states.position
+                self._feedback.actual.velocities = joint_states.velocity
+                self._feedback.actual.time_from_start = rospy.Time.from_sec(time.time()) - time_start
+            self._as.publish_feedback(self._feedback)
```

Comment:
I'm not following, what is the point of this feedback publisher? Where does this go?

Reply:
So since this is an action server, it publishes feedback while the arm is in the process of moving along the trajectory. Essentially actions are services, where there is a request and response, but then while it is being processed another node (rviz) wants to know what is happened during execution. The feedback is subscribed to by rviz to know the current position of the arm. The action definition can be found here:
http://docs.ros.org/en/noetic/api/control_msgs/html/action/FollowJointTrajectory.html